### PR TITLE
Feature/ts new navigationcontroller

### PR DESCRIPTION
**Pull request by alimoncul**
Only android done. Someone should continue this.

### CODE REVIEW

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  set onTransition(callback: (opts?: { controller: Controller; operation: OperationType }) => void) {
+    this._onTransitionCallback = callback;
+  }
+  toString() {
+    return 'NavigationController';
+  }
+  // Use this function to show page or controller without back stack operation.
+  // Show page or controller that exists in history
+  // Call this function from BottomTabBarController
+  show(params?: any) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:76]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    //   operation: NavigationController.OperationType.PUSH
+    // });
+    // TODO: Chnage currentPage as currentController
+    this._onTransitionCallback?.({
+      controller: currentController,
+      targetController: params.controller,
+      operation: OperationType.PUSH
+    });
+  }
+  push(params) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:111]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    this.__isActive && ViewController.deactivateController(this.getCurrentController());
+
+    params.controller.parentController = this;
+    params.controller.isInsideBottomTabBar = this.isInsideBottomTabBar;
+    this.pageIDCollectionInStack[params.controller.pageID] = params.controller;
+    this._childControllers.push(params.controller);
+    this.show(params);
+  }
+  showController(params) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:128]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+      });
+    } else if (params.controller instanceof BottomTabBarController) {
+      params.controller.isInsideBottomTabBar = true;
+      params.controller.show();
+    } else {
+      throw new Error('The controller is not a Page instance or a BottomTabBarController instance!');
+    }
+  }
+
+  present(params) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:145]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    ViewController.setController({
+      controller: params.controller,
+      animation: params.animated,
+      isComingFromPresent: true
+    });
+
+    params.onComplete && params.onComplete();
+  }
+  dismiss(params: any) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:159]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+    Application.currentPage = this.popUpBackPage;
+    ViewController.activateRootController(Application.currentPage);
+    this.popUpBackPage = null;
+    this.popupBackNavigator = null;
+
+    ViewController.deactivateController(this);
+    params.onComplete && params.onComplete();
+  }
+  pop(params) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:175]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    const poppedController = this._childControllers.pop();
+    this.pageIDCollectionInStack[poppedController.pageID] = null;
+    if (!this.__isActive) {
+      return;
+    }
+
+    !params && (params = {});
+    this.popFromHistoryController(poppedController, params);
+  }
+  popTo(params) {
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:189]: params tipini verelim

File: ui/navigationcontroller/navigationcontroller.android.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+  }
+  // TODO: Use getCurrentController for all possible case
+  getCurrentController() {
+    if (this._childControllers.length > 0) {
+      return this._childControllers[this._childControllers.length - 1];
+    }
+    return null;
+  }
+  popFromHistoryController(currentController, params) {
+    const targetController = this._childControllers[this._childControllers.length - 1];
```
Comment (cenkce) [inline ui/navigationcontroller/navigationcontroller.android.ts:220]: parametrelerin tiplerini verelim

**Review by cenkce — CHANGES_REQUESTED** [review]

**Comment by furkanarabaci** [comment]
> Only android done. Someone should continue this.

Taking over.

**Pull request merged**